lib/iris/cube.py

```diff
@@ -3940,7 +3940,15 @@ def __next__(self):
 
         if self._ordered:
             if any(self._mod_requested_dims != list(range(len(cube.shape)))):
-                cube.transpose(self._mod_requested_dims)
+                sliced_dims = []
+                for i in self._requested_dims:
+                    # Convert dim numbers back to dims to match up to indices
+                    # of sliced cube
+                    dim_ref = self._cube.dim_coords[i].name()
```

Comment:
Although I am a little confused about why this passes all the tests but fails for that thing that you have done above.